exporter/newrelicexporter/transformer.go

```diff
@@ -22,124 +22,134 @@ import (
+			resource.Attributes(),
+		),
 	}
 
-	startTime := t.Timestamp(span.StartTime)
+	if !lib.IsNil() {
+		if n := lib.Name(); n != "" {
+			t.ResourceAttributes[instrumentationNameKey] = n
+			if v := lib.Version(); v != "" {
+				t.ResourceAttributes[instrumentationVersionKey] = v
+			}
+		}
+	}
+	return t
+}
 
-	var (
-		traceID          trace.TraceID
-		spanID, parentID trace.SpanID
-	)
+var (
+	emptySpan         telemetry.Span
+	errEmptySpan      = errors.New("empty span")
+	errInvalidSpanID  = errors.New("SpanID is invalid")
+	errInvalidTraceID = errors.New("TraceID is invalid")
+)
 
-	copy(traceID[:], span.GetTraceId())
-	copy(spanID[:], span.GetSpanId())
-	copy(parentID[:], span.GetParentSpanId())
+func (t *traceTransformer) Span(span pdata.Span) (telemetry.Span, error) {
+	if span.IsNil() {
+		return emptySpan, errEmptySpan
+	}
+
+	startTime := pdata.UnixNanoToTime(span.StartTime())
 	sp := telemetry.Span{
-		ID:          spanID.String(),
-		TraceID:     traceID.String(),
-		Name:        t.TruncatableString(span.GetName()),
-		Timestamp:   startTime,
-		Duration:    t.Timestamp(span.EndTime).Sub(startTime),
-		ServiceName: t.ServiceName,
-		Attributes:  t.SpanAttributes(span),
+		// HexString validates the IDs, it will be an empty string if invalid.
+		ID:         span.SpanID().HexString(),
+		TraceID:    span.TraceID().HexString(),
+		ParentID:   span.ParentSpanID().HexString(),
+		Name:       span.Name(),
+		Timestamp:  startTime,
+		Duration:   pdata.UnixNanoToTime(span.EndTime()).Sub(startTime),
+		Attributes: t.SpanAttributes(span),
 	}
 
-	if parentID != emptySpanID {
-		sp.ParentID = parentID.String()
+	if sp.ID == "" {
+		return sp, errInvalidSpanID
+	}
+	if sp.TraceID == "" {
+		return sp, errInvalidTraceID
 	}
 
 	return sp, nil
 }
 
-func (t *transformer) TruncatableString(ts *tracepb.TruncatableString) string {
-	if ts == nil {
-		return ""
-	}
-	return ts.Value
-}
-
-func (t *transformer) SpanAttributes(span *tracepb.Span) map[string]interface{} {
-
-	length := 2
+func (t *traceTransformer) SpanAttributes(span pdata.Span) map[string]interface{} {
 
-	isErr := span.Status != nil && span.Status.Code != 0
-	if isErr {
-		length++
-	}
+	length := 2 + len(t.ResourceAttributes) + span.Attributes().Len()
 
-	if t.Resource != nil {
-		length += len(t.Resource.Labels)
+	var hasStatusCode, hasStatusDesc bool
+	if s := span.Status(); !s.IsNil() {
+		if s.Code() != pdata.StatusCodeUnset {
+			hasStatusCode = true
+			length++
+			if s.Message() != "" {
+				hasStatusDesc = true
+				length++
+			}
+		}
 	}
 
-	if span.Attributes != nil {
-		length += len(span.Attributes.AttributeMap)
+	validSpanKind := span.Kind() != pdata.SpanKindUNSPECIFIED
+	if validSpanKind {
+		length++
 	}
 
 	attrs := make(map[string]interface{}, length)
 
-	// Any existing error attribute will override this.
-	if isErr {
-		attrs["error"] = true
+	if hasStatusCode {
+		code := strings.TrimPrefix(span.Status().Code().String(), "STATUS_CODE_")
```

Comment:
Consider a switch statement for some performance improvement here and for Kind

Reply:
I'm not sure I follow. Each one of these conditional statements is independent. They each have unique clauses that are decoupled from one another.

To replicate the same behavior they would be transformed to a switch statement with one case. Is that what you envisioned?